### ISSUE TITLE
Expose WalkFileDescriptorSet(Visitor, *descriptor.FileDescriptorSet) error?

### ISSUE THREAD

**Issue by peter-edge** [lyft/protoc-gen-star#10]
In my use case, I derive the `*descriptor.FileDescriptorSet` by shelling out to `protoc` internally to my Golang program with `-o tmpFile --include_source_info` and use a cached/versioned protoc. This would not be too complicated, the code is pretty much there via https://github.com/lyft/protoc-gen-star/blob/master/testdata/protoc-gen-example/printer_module.go with smart defaults. Thoughts?

**Comment by rodaine** [comment]
I like the utility of having a visitor pattern for an arbitrary set of file descriptors, and this would also benefit from exposing the logic to convert from a the set to a slice/map of `pgs.Package`. There are a few places where we relying on the protoc-gen-go `generator.Descriptor` types which have a few extra facilities that would make it harder to use the set directly. Reducing the reliance on those would be beneficial (less reliance on PGG internals).

This could also allow for an alternative approach to providing input for the PG* generator. By default, we use stdin (considering the relationship with protoc), but provide a facility to get a CodeGeneratorRequest from [an arbitrary `io.Reader`](https://godoc.org/github.com/lyft/protoc-gen-star#ProtocInput). Allowing for a `FileDescriptorSet` as well would be great for testing.

tl;dr: I'm game!

**Comment by peter-edge** [comment]
Great :-) I generally like the abstraction away from globals, including os.Stdin - enables so many more use cases.

I can send this to you when I have time if you'd like? Or I can leave it to you - let me know!

**Comment by rodaine** [comment]
Feel free to take a stab at it!

**Comment by peter-edge** [comment]
Note re #7 - it's very useful to have some library-native handling around options, there's a ton of commonality. Would be great to have a full-FileDescriptorSet walker here, for use either as a plugin with `protoc`, or within separate Golang code :-)

**Comment by rodaine** [comment]
By option are you referring to the proto extensions or in terms of the flexibility of manipulating unmarshaled PB structs with the utilities here?

If it's the former, you can access extension values via [Extension](https://godoc.org/github.com/lyft/protoc-gen-star#Entity) on all Entity types. You can see an example use of this in [protoc-gen-validate](https://github.com/lyft/protoc-gen-validate/blob/master/checker.go#L31).

If it's the latter, I am in total agreement. Having an exported function that converts `FileDescriptorSet` into the PG* entities would be powerful (like your use case) and having a Walk that takes a set (internally running that conversion) would be great too.

**Comment by clintberry** [comment]
The Extension function takes a proto object as an argument? Does that mean I have to pre-build my proto extension in Go before I can access extension values from the proto I am generating off of?

**Comment by rodaine** [comment]
This is a limitation of the protobuf library. It won't know how to deserialize the extensions if you don't provide it the descriptor. The implementation should point to why this is so problematic: https://github.com/lyft/protoc-gen-star/blob/master/extension.go#L32

**Comment by clintberry** [comment]
In my case, the extensions definition is in the proto file I am loading as an import. In that case, it should already be aware. Do I need to find it in the tree first to pass it as a proto object?

**Comment by clintberry** [comment]
In fact, if anyone is using extensions they have to include the definitions in their proto file, or as an imported file, right? So protoc should have already parsed them. Maybe I am thinking about this wrong.

**Comment by clintberry** [comment]
If I was to submit a pull request putting a couple small helper functions on the field object like so: `pgs.Field.ExtensionString(int32 id)` where `id` is the extension id, would that be accepted? Here is the code I would use to model the functions: https://github.com/moul/protoc-gen-gotemplate/pull/93/commits/6765ff04ebc9384dac43486bf4e2d1ec4fef9c47#diff-6f9eba605ec12ce93b6e424662b73469R253

**Comment by clintberry** [comment]
Cool, I will give it a go.

**Comment by clintberry** [comment]
Here is the start. https://github.com/lyft/protoc-gen-star/pull/23

It compiles fine but returns weird errors when I try to call it. I could be in over my head. If you see anything obvious then any help is appreciated.

**Comment by peter-edge** [comment]
Closing this for now, re-open if someone pursues this further!
